In [1]:
import re

## Functions

In [3]:
def prep_lines(lines):
    '''
    Remove Fortran-specific code, and remove \ n from reading in as strings. 
    '''
    for i in range(len(lines)):
        lines[i]=lines[i].replace('     * ', '')
        lines[i]=lines[i].replace(' \n', '')
        lines[i]=lines[i].replace('\n', '')
    return lines


def read_fortran_ODEs(file, cons = False):
    '''
    Read fortran ODEs and prep for translation.\n
        - file = file name\n
        - returns: dict
    '''

    ## read fortran ODEs from given file
    with open(file, 'r') as f:
        lines = f.readlines()

    if cons == True:
        lines = lines[16:93]
    if cons == False:
        ## select non-conserved species
        lines = lines[123:]

    ## remove unnecessary fortran code
    lines = prep_lines(lines)

    ODEs = stick_fortran(lines)

    return ODEs


def stick_fortran(lines):
    '''
    Make dictionary:\n
        - entry per species\n
        - stick together separate fortran strings (result: full ODE in 1 string) \n
        \n
    This is possible by searching for 'C' in Fortran, indicating commented code.
    '''

    ODEs = dict()

    for i, line in enumerate(lines):
        if len(line) != 0 and line[0] == 'C':
            ODEs[line[2:]] = str('')
            print('i=',i, end = '\r')
            start = i + 1

            for j in range(start,start+1000):
                if j >= len(lines):
                    break
                if len(lines[j]) != 0 and lines[j][0] == 'C':
                    stop = j
                    break
                
            for j in range(i+1,stop):
                ODEs[line[2:]] += lines[j]

            i = stop +1

    count = 0
    for key in ODEs:
        if key == '':
            count += 1

    while count > 0:
        del ODEs['']
        count -= 1

    return ODEs



def fortran_to_python(ODEs, spec, cons = False):
    '''
    Translate fortran ODE to python ODE. \n\n

    Note: NOT fractional abundances, but in units of cm^-3.
    '''
    
    print(spec)

    if isinstance(ODEs[spec],str):
        ## split string
        ODEs[spec] = ODEs[spec].split(' ')

    ## remove empty strings
    to_remove = ''
    while to_remove in ODEs[spec]:
        ODEs[spec].remove('')

    ## remove other fortran commands
    while 'END' in ODEs[spec]:
        ODEs[spec].remove('END')
    while 'RETURN' in ODEs[spec]:
        ODEs[spec].remove('RETURN')

    
    ## Translate

    ## Conserved species
    if cons == True:
        ## replace brackets: ( --> [
        print(len(ODEs[spec]))
        for i in range(len(ODEs[spec])):
            
            ODEs[spec][i] = ODEs[spec][i].replace('(', '[')
            ODEs[spec][i] = ODEs[spec][i].replace(')', ']')
            ODEs[spec][i] = ODEs[spec][i].replace('[+', '(+')
            ODEs[spec][i] = ODEs[spec][i].replace(']]', '])')

    ## Non-conserved species
    if cons == False:
        ## replace brackets: ( --> [
        for i in range(len(ODEs[spec])):
            if ODEs[spec][i][0] == 'F' or ODEs[spec][i][0] == 'D':
                ODEs[spec][i] = ODEs[spec][i].replace('(', '[')
                ODEs[spec][i] = ODEs[spec][i].replace(')', ']')

        ## replace final line --> sets the change in number density of species
        last = ODEs[spec][-1]
        while not last[0] == 'Y':
            ODEs[spec].remove(last)
            last = ODEs[spec][-1]
        for i in range(5,len(last)):
            if last[i] == ')':
                stop = i
                break
        nb = last[5:stop]

        if not spec == 'H':
            ODEs[spec][-1] = 'YDOT['+nb+']=F-(D*Y['+nb+'])'
        if spec == 'H':
            ODEs[spec][-1] = 'YDOT['+nb+']=(F+HLOSS)-(D*Y['+nb+'])'

            
    return ODEs


def parse_line(parsed, cons = False):
    '''
    Parse line to print is too long (Python doesn't like that much).
    '''
    
    if cons == False:
        line = parsed[-1]
        
        if len(line) > 1000:
            if cons == False:
                for i in range(1000):
                    n=i+1000
                    if n < len(line):
                        if line[n] == '+' or line[n] == '-':
                            break
                    else:
                        break
                parsed[-1] = line[0:n]
            
                parsed.append(line[0]+'='+line[0]+line[n:])            

        else: 
            print('No parsing need')
    
    return 

def minus_one(match):
    '''
    Helper function that is used in the regular expression substitution.
    '''
    # Extract the mached string from the matched object
    match = match.group(0)
    # remove square brackets
    match = match[1:-1]
    # Cast to integer
    match = int(match)    
    # Minus one
    match -= 1
    # Add brackets again
    match = f'[{match}]'

    
    return match
    

## Read in file & prep

In [4]:
file = '/lhome/silkem/CHEM/src-IP-AP-HNR/code/acodes.f'

with open(file, 'r') as f:
    lines = f.readlines()

ODEs = read_fortran_ODEs(file)


In [5]:
ODEs['C11+']

'      F=0.+K(1517)*Y(10)*Y(458)*HNR+K(1518)*Y(10)*Y(455)*HNR+K(2066)*Y(11)*Y(453)*HNR+K(2068)*Y(11)*Y(459)*HNR+K(5727)*Y(465)+K(6042)*Y(10)*Y(451)*HNR+K(6202)*Y(465)+K(6538)*Y(465)      D=0.+K(994)*X(1)*HNR+K(995)*X(1)*HNR+K(996)*X(1)*HNR+K(997)*X(1)*HNR      YDOT(466)=F-(D*Y(466))      RETURN      END'

## Translate

### Non-conserved species

In [6]:
for key in ODEs:
    ODEs = fortran_to_python(ODEs, key)

H
H+
H-
H2+
H3+
He
He+
HeH+
C-
C+
C
CH-
CH+
CH
CH2+
N+
CH2
N
CH3
NH
NH+
CH3+
O+
NH2+
CH4+
O-
NH2
O
CH4
OH-
OH
OH+
CH5+
NH3+
NH3
NH4+
H2O
H2O+
F
H3O+
F+
HF
HF+
H2F+
Na
Na+
C2-
C2+
Mg
Mg+
C2
C2H-
C2H+
C2H
CN
C2H2
C2H2+
CN-
CN+
HCN
C2H3
C2H3+
HCN+
HNC
N2
C2H4+
Si+
H2CN
H2NC+
N2+
Si
CO
CO+
C2H4
HCNH+
C2H5+
HCO
SiH+
HCO+
CH2NH
SiH
C2H5
HOC+
N2H+
CH2NH2+
SiH2
CH3CH3+
NO
H2CO
CH4N+
H2CO+
CH3CH3
SiH2+
NO+
P+
P
H3CO+
HNO+
SiH3+
HNO
CF+
C2H7+
SiH3
O2
SiH4
PH
S
H2NO+
CH3OH
SiH4+
O2+
S-
O2-
PH+
CH3OH+
S+
CH3OH2+
O2H
PH2+
HS
HS+
PH2
O2H+
SiH5+
H2O2
H2S+
H2S
PH3+
H3S+
Cl
Cl+
C3+
HCl
C3-
C3
HCl+
H2Cl+
C3H-
C3H+
C3H
C2N
C3H2+
C2N+
H2CCC
CNC+
C3H2
C2NH+
C3H3+
CH2CCH
CH2CCH+
SiC
CH2CCH2
SiC+
C2O
C3H4+
C2O+
CH2CN
CH3CCH
CH2CN+
C3H5+
HCSi
HCSi+
CH3CN+
HC2O+
CH3CN
C3H6+
SiN
CH3CNH+
NH2CN
CH3CHCH2
SiN+
SiCH2+
SiCH2
CNO
CH2CO+
OCN
OCN+
CH2CO
CP+
HNCO
HCNO+
HNCO+
HONC
SiCH3+
HCNO
CP
C3H7+
NH2CNH+
HNSi+
HOCN
HNSi
CH3CO+
HOCN+
SiCH3
HONC+
CS+
HCNOH+
N2O+
CS
SiCH4+
CO2+
H2OCN+
HNCOH+
SiNH2+
HCP
SiO+
CH3CHO
SiO
H

In [7]:
## Parsing the ODEs of the non-conserved specs
for spec in ODEs:
    print(spec)

    i = 0
    while i < len(ODEs[spec]):

        if len(ODEs[spec][i]) > 1000:
            parsed = [ODEs[spec][i]]
            while len(parsed[-1]) > 1000:
                parse_line(parsed)

            ODEs[spec] = ODEs[spec][0:i] + parsed + ODEs[spec][i+1:]

            i = i + len(parsed)
        
        else: i += 1

H
H+
H-
H2+
H3+
He
He+
HeH+
C-
C+
C
CH-
CH+
CH
CH2+
N+
CH2
N
CH3
NH
NH+
CH3+
O+
NH2+
CH4+
O-
NH2
O
CH4
OH-
OH
OH+
CH5+
NH3+
NH3
NH4+
H2O
H2O+
F
H3O+
F+
HF
HF+
H2F+
Na
Na+
C2-
C2+
Mg
Mg+
C2
C2H-
C2H+
C2H
CN
C2H2
C2H2+
CN-
CN+
HCN
C2H3
C2H3+
HCN+
HNC
N2
C2H4+
Si+
H2CN
H2NC+
N2+
Si
CO
CO+
C2H4
HCNH+
C2H5+
HCO
SiH+
HCO+
CH2NH
SiH
C2H5
HOC+
N2H+
CH2NH2+
SiH2
CH3CH3+
NO
H2CO
CH4N+
H2CO+
CH3CH3
SiH2+
NO+
P+
P
H3CO+
HNO+
SiH3+
HNO
CF+
C2H7+
SiH3
O2
SiH4
PH
S
H2NO+
CH3OH
SiH4+
O2+
S-
O2-
PH+
CH3OH+
S+
CH3OH2+
O2H
PH2+
HS
HS+
PH2
O2H+
SiH5+
H2O2
H2S+
H2S
PH3+
H3S+
Cl
Cl+
C3+
HCl
C3-
C3
HCl+
H2Cl+
C3H-
C3H+
C3H
C2N
C3H2+
C2N+
H2CCC
CNC+
C3H2
C2NH+
C3H3+
CH2CCH
CH2CCH+
SiC
CH2CCH2
SiC+
C2O
C3H4+
C2O+
CH2CN
CH3CCH
CH2CN+
C3H5+
HCSi
HCSi+
CH3CN+
HC2O+
CH3CN
C3H6+
SiN
CH3CNH+
NH2CN
CH3CHCH2
SiN+
SiCH2+
SiCH2
CNO
CH2CO+
OCN
OCN+
CH2CO
CP+
HNCO
HCNO+
HNCO+
HONC
SiCH3+
HCNO
CP
C3H7+
NH2CNH+
HNSi+
HOCN
HNSi
CH3CO+
HOCN+
SiCH3
HONC+
CS+
HCNOH+
N2O+
CS
SiCH4+
CO2+
H2OCN+
HNCOH+
SiNH2+
HCP
SiO+
CH3CHO
SiO
H

### Conserved species

In [8]:
ODEs_cons = read_fortran_ODEs(file, True)

In [9]:
for spec in ODEs_cons:
    ODEs_cons = fortran_to_python(ODEs_cons, spec, cons = True)

e-
1
H2
1


## Write out in as python function

In [10]:
file = 'src/odes.py'

## Y = number density of unconserved species
## X = number density of conserved species
## K = reaction rate
## ACCR = loss of H due to dust interaction, kind of rate

with open(file, 'w') as f:
    f.write('import numpy as np\n')
    f.write("\n'''\nIMPORTANT NOTE:\nThe content of this function is translated from Fortran 77. \n \nWritten by Silke Maes, April 2023.\nOriginal Fortran code by rate10odecsT.pl on Sun, Jul 21 2019, \nusing ratefile: rate16_IP_2330K_AP_6000K.rates, from http://udfa.ajmarkwick.net/index.php. \n")
    f.write("\n------\n")
    f.write("\nY     = number density of the non-conserved species")
    f.write("\nYDOT  = change in number density of the conserved species")
    f.write("\nX     = number density of the conserved species during calculation")
    f.write("\nK     = reaction rates, calculated in xxx")
    f.write("\nACCR  = H accretion onto dust, calculated in main.py from input physics")
    f.write("\nTOTAL = initial number density of conserved species")
    f.write("\nHNR   = input density (ρ)")
    f.write("\n'''")
    f.write('\ndef ODE(t, Y, YDOT, X, TOTAL, K, HNR, ACCR):\n')
    f.write('\n\t## Conserved species\n')
    for key in ODEs_cons:
        f.write('\n\t## '+str(key)+'\n')
        for line in ODEs_cons[key]:
            f.write('\t'+line+'\n')
    f.write('\n\n\t## Non-conserved species\n')
    f.write('\tHLOSS=-ACCR*Y[1]\n')
    for key in ODEs:
        f.write('\n\t## '+str(key)+'\n')
        for line in ODEs[key]:
            f.write('\t'+line+'\n')
    f.write('\n\treturn YDOT')

# Read the original ODE file
with open('src/odes.py', 'r') as file:
    content = file.read()

# Substitute every number between square brackets with that number minus one
content_new = re.sub('\[\d+\]', minus_one, content)

# Write the new ODE file
with open('src/odes.py', 'w') as file:
    file.write(content_new)

In [47]:
file = 'src/ode/odes_test1_tmp.py'

## Y = number density of unconserved species
## X = number density of conserved species
## K = reaction rate
## ACCR = loss of H due to dust interaction, kind of rate

with open(file, 'w') as f:
    f.write('import numpy as np\n')
    f.write("\n'''\nIMPORTANT NOTE:\nThe content of this function is translated from Fortran 77. \nHence, indexing starts from 1. At position 0 there is a dummy species.\n \nWritten by Silke Maes, April 2023.\nOriginal Fortran code by rate10odecsT.pl on Sun, Jul 21 2019, \nusing ratefile: rate16_IP_10000K.rates, from http://udfa.ajmarkwick.net/index.php. \n")
    f.write("\n------\n")
    f.write("\nt    = time")
    f.write("\nY    = abundances of the non-conserved species")
    f.write("\nargs = list of arguments of the function:")
    f.write("\n     [0]: TOTAL = ??")
    f.write("\n     [1]: K     = reaction rates, calculated in xxx")
    f.write("\n     [2]: HNR   = input density (ρ)")
    f.write("\n     [3]: ACCR  = H accretion onto dust, calculated in main.py from input physics")   
    f.write("\nX    = abundances of the conserved species")
    f.write("\n'''")
    f.write('\ndef ODE(t,Y, args):\n')
    f.write('\n\tTOTAL = args[1]')
    f.write('\n\tK     = args[2]')
    f.write('\n\tHNR   = args[3]')
    f.write('\n\tACCR  = args[4]\n')
    f.write('\n\tX     = np.empty('+str(len(ODEs_cons))+')')
    f.write('\n\tYDOT  = np.empty('+str(len(ODEs))+')\n')
    f.write('\n\n\t## Conserved species\n')
    for key in ODEs_cons:
        f.write('\n\t## '+str(key)+'\n')
        for line in ODEs_cons[key]:
            f.write('\t'+line+'\n')
    f.write('\n\n\t## Non-conserved species\n')
    f.write('\tHLOSS=-ACCR*Y[1]\n')

    for key in ODEs:
        f.write('\n\t## '+str(key)+'\n')
        for line in ODEs[key]:
            f.write('\t'+line+'\n')
    f.write('\n\treturn YDOT')

# Read the original ODE file
with open('src/ode/odes_test1_tmp.py', 'r') as file:
    content = file.read()

# Substitute every number between square brackets with that number minus one
content_new = re.sub('\[\d+\]', minus_one, content)

# Write the new ODE file
with open('./src/ode/odes_test1.py', 'w') as file:
    file.write(content_new)

In [48]:
file = '/src/ode/odes_test2_tmp.py'

## Y = number density of unconserved species
## X = number density of conserved species
## K = reaction rate
## ACCR = loss of H due to dust interaction, kind of rate

with open(file, 'w') as f:
    f.write('import numpy as np\n')
    f.write("\n'''\nIMPORTANT NOTE:\nThe content of this function is translated from Fortran 77. \nHence, indexing starts from 1. At position 0 there is a dummy species.\n \nWritten by Silke Maes, April 2023.\nOriginal Fortran code by rate10odecsT.pl on Sun, Jul 21 2019, \nusing ratefile: rate16_IP_10000K.rates, from http://udfa.ajmarkwick.net/index.php. \n")
    f.write("\n------\n")
    f.write("\nt    = time")
    f.write("\nY    = abundances of the non-conserved species")
    f.write("\nX    = abundances of the conserved species")
    f.write("\nYDOT = change in abundances of the conserved species")
    f.write("\nargs = list of arguments of the function:")
    f.write("\n     [0]: TOTAL = ??")
    f.write("\n     [1]: K     = reaction rates, calculated in xxx")
    f.write("\n     [2]: HNR   = input density (ρ)")
    f.write("\n     [3]: ACCR  = H accretion onto dust, calculated in main.py from input physics")   

    f.write("\n'''")
    f.write('\ndef ODE(t,Y, X, YDOT, args):\n')
    f.write('\n\tTOTAL = args[1]')
    f.write('\n\tK     = args[2]')
    f.write('\n\tHNR   = args[3]')
    f.write('\n\tACCR  = args[4]\n')
    f.write('\n\n\t## Conserved species\n')
    for key in ODEs_cons:
        f.write('\n\t## '+str(key)+'\n')
        for line in ODEs_cons[key]:
            f.write('\t'+line+'\n')
    f.write('\n\n\t## Non-conserved species\n')
    f.write('\tHLOSS=-ACCR*Y[1]\n')

    for key in ODEs:
        f.write('\n\t## '+str(key)+'\n')
        for line in ODEs[key]:
            f.write('\t'+line+'\n')
    f.write('\n\treturn YDOT')

# Read the original ODE file
with open('scr/ode/odes_test2_tmp.py', 'r') as file:
    content = file.read()

# Substitute every number between square brackets with that number minus one
content_new = re.sub('\[\d+\]', minus_one, content)

# Write the new ODE file
with open('./scr/ode/odes_test2.py', 'w') as file:
    file.write(content_new)

FileNotFoundError: [Errno 2] No such file or directory: '/src/ode/odes_test2_tmp.py'

'''
ODEs for the conserved species, manually parsed --> plug in final file
'''


## e-
T = 0.+Y[1]-Y[2]+Y[3]+Y[4]+Y[6]+Y[7]-Y[8]+Y[9]-Y[11]+Y[12]+Y[14]+Y[15]+Y[20]+Y[21]+Y[22]+Y[23]+Y[24]-Y[25]-Y[29]+Y[31]+Y[32]+Y[33]+Y[35]+Y[37]+Y[39]+Y[40]+Y[42]+Y[43]+Y[45]-Y[46]+Y[47]+Y[49]-Y[51]+Y[52]+Y[56]-Y[57]+Y[58]+Y[61]+Y[62]+Y[65]+Y[66]+Y[68]+Y[69]+Y[72]+Y[74]+Y[75]+Y[77]+Y[78]+Y[82]+Y[83]+Y[84]+Y[86]+Y[89]+Y[90]+Y[92]+Y[93]+Y[94]+Y[96]+Y[97]+Y[98]+Y[100]+Y[101]+Y[107]+Y[109]+Y[110]-Y[111]-Y[112]+Y[113]+Y[114]+Y[115]+Y[116]+Y[118]+Y[120]+Y[122]+Y[123]+Y[125]+Y[127]+Y[128]+Y[130]+Y[131]-Y[133]+Y[135]+Y[136]-Y[137]+Y[138]+Y[141]+Y[142]+Y[144]+Y[146]+Y[147]+Y[149]+Y[152]+Y[154]+Y[155]+Y[158]+Y[159]+Y[161]+Y[162]+Y[163]+Y[165]+Y[167]+Y[170]+Y[171]+Y[174]+Y[176]+Y[178]+Y[180]+Y[181]+Y[183]+Y[186]+Y[187]+Y[188]+Y[191]+Y[192]+Y[194]+Y[195]+Y[196]+Y[197]+Y[199]+Y[200]+Y[201]+Y[202]+Y[203]+Y[205]+Y[208]+Y[209]+Y[211]+Y[212]+Y[214]+Y[215]+Y[216]+Y[217]+Y[220]+Y[221]+Y[222]+Y[223]+Y[228]+Y[229]+Y[232]+Y[234]+Y[235]+Y[236]+Y[238]+Y[239]+Y[240]+Y[241]+Y[242]+Y[244]+Y[245]+Y[246]+Y[247]+Y[248]+Y[249]+Y[251]+Y[252]-Y[253]
T = T+Y[254]+Y[258]+Y[259]+Y[260]+Y[261]+Y[262]+Y[263]-Y[265]+Y[266]+Y[267]+Y[270]-Y[271]+Y[272]+Y[276]+Y[277]+Y[282]+Y[284]+Y[285]+Y[286]+Y[287]+Y[288]+Y[289]+Y[292]+Y[293]+Y[294]+Y[295]+Y[296]+Y[298]+Y[300]+Y[302]+Y[305]+Y[306]+Y[307]+Y[308]+Y[310]+Y[311]+Y[312]+Y[314]+Y[316]+Y[317]+Y[318]+Y[320]+Y[321]+Y[322]+Y[323]+Y[326]-Y[328]+Y[329]+Y[332]+Y[333]+Y[334]+Y[335]+Y[336]+Y[337]+Y[339]-Y[340]+Y[341]+Y[342]+Y[345]+Y[346]+Y[350]+Y[351]+Y[352]+Y[354]+Y[355]+Y[357]+Y[358]+Y[359]+Y[360]+Y[363]+Y[364]+Y[365]+Y[366]+Y[368]+Y[369]+Y[371]+Y[373]+Y[374]-Y[375]+Y[376]+Y[379]-Y[380]+Y[382]-Y[384]+Y[385]+Y[386]+Y[388]+Y[389]+Y[390]+Y[391]+Y[393]+Y[394]+Y[395]+Y[397]+Y[399]+Y[400]+Y[402]+Y[403]+Y[404]+Y[405]-Y[407]-Y[408]+Y[409]+Y[411]+Y[413]+Y[414]+Y[417]+Y[418]-Y[419]+Y[421]+Y[423]-Y[424]+Y[427]+Y[428]+Y[430]+Y[431]+Y[432]+Y[433]+Y[434]+Y[435]+Y[437]-Y[438]+Y[439]-Y[440]+Y[442]+Y[444]+Y[445]+Y[447]+Y[448]-Y[449]+Y[451]+Y[452]-Y[453]+Y[456]+Y[458]+Y[459]+Y[460]+Y[462]+Y[463]+Y[465]
X[0]=0.+TOTAL[0]+T

## H2
T = 0.+Y[0]+Y[1]+Y[2]+2*Y[3]+3*Y[4]+Y[7]+Y[11]+Y[12]+Y[13]+2*Y[14]+2*Y[16]+3*Y[18]+Y[19]+Y[20]+3*Y[21]+2*Y[23]+4*Y[24]+2*Y[26]+4*Y[28]+Y[29]+Y[30]+Y[31]+5*Y[32]+3*Y[33]+3*Y[34]+4*Y[35]+2*Y[36]+2*Y[37]+3*Y[39]+Y[41]+Y[42]+2*Y[43]+Y[51]+Y[52]+Y[53]+2*Y[55]+2*Y[56]+Y[59]+3*Y[60]+3*Y[61]+Y[62]+Y[63]+4*Y[65]+2*Y[67]+2*Y[68]+4*Y[73]+2*Y[74]+5*Y[75]+Y[76]+Y[77]+Y[78]+3*Y[79]+Y[80]+5*Y[81]+Y[82]+Y[83]+4*Y[84]+2*Y[85]+6*Y[86]+2*Y[88]+4*Y[89]+2*Y[90]+6*Y[91]+2*Y[92]+3*Y[96]+Y[97]+3*Y[98]+Y[99]+7*Y[101]+3*Y[102]+4*Y[104]+Y[105]+2*Y[107]+4*Y[108]+4*Y[109]+Y[113]+4*Y[114]+5*Y[116]+Y[117]+2*Y[118]+Y[119]+Y[120]+2*Y[121]+Y[122]+5*Y[123]+2*Y[124]+2*Y[125]+2*Y[126]+3*Y[127]+3*Y[128]+Y[132]+Y[135]+2*Y[136]+Y[137]+Y[138]+Y[139]+2*Y[141]+2*Y[143]+2*Y[145]+Y[146]+3*Y[147]+3*Y[148]+3*Y[149]+4*Y[151]+4*Y[154]+2*Y[156]+4*Y[157]+2*Y[158]+5*Y[159]+Y[160]+Y[161]+3*Y[162]+Y[163]+3*Y[164]+6*Y[165]+4*Y[167]+2*Y[168]+6*Y[169]+2*Y[171]+2*Y[172]+2*Y[174]+2*Y[177]+Y[179]+Y[180]+Y[181]+Y[182]+3*Y[183]+Y[184]+7*Y[186]+3*Y[187]+Y[188]+Y[189]+Y[190]
T = T+3*Y[191]+Y[192]+3*Y[193]+Y[194]+2*Y[196]+4*Y[199]+2*Y[201]+2*Y[202]+2*Y[203]+Y[204]+4*Y[206]+2*Y[208]+4*Y[209]+2*Y[211]+Y[212]+5*Y[214]+Y[215]+Y[216]+Y[219]+2*Y[220]+Y[221]+Y[222]+2*Y[224]+2*Y[225]+3*Y[227]+2*Y[228]+6*Y[229]+6*Y[231]+2*Y[232]+2*Y[233]+3*Y[234]+2*Y[235]+6*Y[236]+6*Y[237]+Y[239]+7*Y[240]+2*Y[242]+3*Y[244]+Y[247]+4*Y[248]+3*Y[249]+7*Y[251]+3*Y[252]+Y[254]+Y[255]+3*Y[260]+Y[261]+2*Y[262]+2*Y[263]+Y[264]+Y[265]+Y[266]+2*Y[267]+2*Y[269]+Y[272]+3*Y[273]+Y[274]+3*Y[277]+Y[278]+4*Y[280]+2*Y[284]+4*Y[285]+Y[288]+5*Y[289]+Y[290]+3*Y[291]+3*Y[292]+Y[293]+Y[294]+2*Y[295]+4*Y[296]+6*Y[297]+2*Y[298]+2*Y[299]+Y[302]+5*Y[304]+3*Y[305]+7*Y[307]+3*Y[308]+Y[309]+Y[310]+6*Y[312]+Y[316]+2*Y[317]+6*Y[318]+6*Y[319]+3*Y[320]+7*Y[321]+3*Y[322]+4*Y[323]+4*Y[324]+4*Y[333]+Y[334]+Y[335]+Y[336]+Y[337]+Y[338]+5*Y[339]+Y[340]+2*Y[341]+2*Y[343]+Y[345]+3*Y[346]+4*Y[347]+4*Y[351]+2*Y[354]+Y[356]+Y[357]+Y[358]+3*Y[359]+Y[360]+3*Y[361]+Y[362]+5*Y[363]+4*Y[364]+2*Y[365]+2*Y[366]+2*Y[367]+3*Y[368]+3*Y[369]+Y[371]+Y[374]+Y[378]
T = T+Y[379]+Y[380]+2*Y[381]+2*Y[382]+Y[386]+Y[387]+3*Y[388]+4*Y[390]+2*Y[391]+3*Y[393]+Y[394]+5*Y[395]+6*Y[396]+6*Y[397]+7*Y[399]+Y[402]+Y[404]+Y[408]+Y[409]+Y[410]+2*Y[411]+2*Y[412]+3*Y[413]+4*Y[414]+4*Y[415]+3*Y[416]+5*Y[417]+4*Y[418]+Y[422]+Y[423]+Y[424]+2*Y[425]+2*Y[428]+Y[429]+Y[430]+3*Y[431]+4*Y[432]+2*Y[433]+5*Y[434]+3*Y[435]+Y[439]+Y[440]+Y[441]+2*Y[442]+2*Y[443]+3*Y[444]+4*Y[445]+3*Y[446]+5*Y[447]+4*Y[448]+Y[452]+Y[453]+Y[454]+2*Y[457]+2*Y[458]+Y[459]+3*Y[460]+Y[461]+2*Y[462]+3*Y[463]
X[1]=0.+TOTAL[1]-0.5*T